<div style="text-align: center; padding: 25px 20px; border-bottom: 4px solid #0D9488; margin-bottom: 35px; background: linear-gradient(135deg, #f0f9f8 0%, #ffffff 100%); border-radius: 8px;">



**<h1><font color="#123A63">SWARM INTELLIGENCE LAB</font></h1>**

**<h3 style="color: #333; margin: 8px 0 0 0; font-weight: 600; font-size: 16px;">Lab 05:** ARTIFICIAL BEE COLONY</h3>


<hr style="border: none; border-top: 2px solid #0D9488; margin: 18px 0;">

| **Name** | **Aiman Imtiaz Ranjha** |
|---|---|
| **Enrollment No** | **01-136232-058** |
| **Program** | **BS(AI)-7** |
| **Instructor** | **Sir Ahsan Javed** |
| **Department** | **Computer Science, Bahria University** |


</div>

In [5]:
import random
random.seed(58) # Setting my own enrolllment random seed as instructed for unique results

def f(x):
    return (x - 7) ** 2 + 4 # The objective function we are trying to minimize

def fitness(x):
    return 1 / (1 + f(x)) # Higher fitness value means we are closer to the true minimum


# Initializing 5 food sources randomly across the continuous search space [0, 14]
sources = [random.uniform(0, 14) for _ in range(5)]

trials = [0 for _ in range(5)]       # Each food source needs its own separate counter to track failed improvements
limit = 8                            # Threshold: if a source doesn't improve after 8 tries, the bee abandons it
iterations = 40                      # Total number of iterations the swarm will run


for it in range(iterations):
    # Employed Bee Phase
    for i in range(5):
        k = random.choice([j for j in range(5) if j != i]) # Pick a different random food source to compare/interact with

        # Move toward or away from source k to explore the local neighborhood
        candidate = sources[i] + random.uniform(-1, 1) * (sources[i] - sources[k])

        # A source only counts as improved if the new spot is strictly better
        if fitness(candidate) > fitness(sources[i]):
            sources[i], trials[i] = candidate, 0      # Keep the new better position and reset the trial counter to 0
        else:
            trials[i] = trials[i] + 1                 # No improvement made, so increment this specific source's failure count

    # Scout Bee Phase
    for i in range(5):
        # If a food source has been stagnant past our limit, abandon it for a fresh random location
        if trials[i] > limit:
            sources[i] = random.uniform(0, 14)        # Abandon stale source and explore somewhere completely new at random
            trials[i] = 0                             # Reset trial counter for the newly discovered scout location

best_x = min(sources, key=f)
print(f"Finished Best x found: {best_x:.4f} | Minimum f(x): {f(best_x):.4f}\n")

Finished Best x found: 7.0000 | Minimum f(x): 4.0000




# **<font color="#0D9488">Experiment 1: Changing the Abandonment Limit (limit = 2** & **limit = 20**)</font>


In [2]:
# EXPERIMENT 1A: Low Limit (limit = 2)
random.seed(58)
limit_low = 2
sources_1 = [random.uniform(0, 14) for _ in range(5)]
trials_1 = [0 for _ in range(5)]

for it in range(40):
    for i in range(5):
        k = random.choice([j for j in range(5) if j != i])
        candidate = sources_1[i] + random.uniform(-1, 1) * (sources_1[i] - sources_1[k])
        if fitness(candidate) > fitness(sources_1[i]):
            sources_1[i], trials_1[i] = candidate, 0
        else:
            trials_1[i] += 1
    for i in range(5):
        if trials_1[i] > limit_low:
            sources_1[i] = random.uniform(0, 14)
            trials_1[i] = 0

print(f"Low Limit (2) Best x: {min(sources_1, key=f):.4f}")


# EXPERIMENT 1B: High Limit (limit = 20)
random.seed(58)
limit_high = 20
sources_2 = [random.uniform(0, 14) for _ in range(5)]
trials_2 = [0 for _ in range(5)]

for it in range(40):
    for i in range(5):
        k = random.choice([j for j in range(5) if j != i])
        candidate = sources_2[i] + random.uniform(-1, 1) * (sources_2[i] - sources_2[k])
        if fitness(candidate) > fitness(sources_2[i]):
            sources_2[i], trials_2[i] = candidate, 0
        else:
            trials_2[i] += 1
    for i in range(5):
        if trials_2[i] > limit_high:
            sources_2[i] = random.uniform(0, 14)
            trials_2[i] = 0

print(f"High Limit (20) Best x: {min(sources_2, key=f):.4f}")

Low Limit (2) Best x: 6.7352
High Limit (20) Best x: 7.0000


# **<font color="#0D9488">Experiment 2: Changing the Number of Food Sources (sources = 3 , 10**) </font>


In [3]:
# EXPERIMENT 2A: Fewer Sources (3 sources)
random.seed(58)
n_sources_3 = 3
sources_3 = [random.uniform(0, 14) for _ in range(n_sources_3)]
trials_3 = [0 for _ in range(n_sources_3)]

for it in range(40):
    for i in range(n_sources_3):
        k = random.choice([j for j in range(n_sources_3) if j != i])
        candidate = sources_3[i] + random.uniform(-1, 1) * (sources_3[i] - sources_3[k])
        if fitness(candidate) > fitness(sources_3[i]):
            sources_3[i], trials_3[i] = candidate, 0
        else:
            trials_3[i] += 1
    for i in range(n_sources_3):
        if trials_3[i] > 8:
            sources_3[i] = random.uniform(0, 14)
            trials_3[i] = 0

print(f"3 Sources Best x: {min(sources_3, key=f):.4f}")


# EXPERIMENT 2B: More Sources (10 sources)
random.seed(58)
n_sources_10 = 10
sources_10 = [random.uniform(0, 14) for _ in range(n_sources_10)]
trials_10 = [0 for _ in range(n_sources_10)]

for it in range(40):
    for i in range(n_sources_10):
        k = random.choice([j for j in range(n_sources_10) if j != i])
        candidate = sources_10[i] + random.uniform(-1, 1) * (sources_10[i] - sources_10[k])
        if fitness(candidate) > fitness(sources_10[i]):
            sources_10[i], trials_10[i] = candidate, 0
        else:
            trials_10[i] += 1
    for i in range(n_sources_10):
        if trials_10[i] > 8:
            sources_10[i] = random.uniform(0, 14)
            trials_10[i] = 0

print(f"10 Sources Best x: {min(sources_10, key=f):.4f}")

3 Sources Best x: 7.0000
10 Sources Best x: 7.0000


# <font color="#0D9488"> **Lab 5 Results & Observations Summary** </font>
#  **Execution (Limit 8, 5 Sources):**
 Successfully converged to the global minimum at **x = 7.0000** with a minimum function value of **4.0000**.
# **Limit Experiment Observations:**
  * Setting a low limit limit = 2 made the swarm restless causing it to abandon positions too early and land slightly off target (**x = 6.7352**).
  * Setting a high limit limit = 20 maintained patience and successfully reached **x = 7.0000**.
#  **Source Count Experiment Observations:**
 Both 3 sources and 10 sources successfully navigated the landscape to find **x = 7.0000**, showing that the ABC algorithm is robust across different population sizes on this landscape.